### ISSUE TITLE
Formie 1.3.21 breaks form submissions

### ISSUE THREAD

**Issue by thijskaspers** [verbb/formie#378]
**Description**

We updated from 1.3.18 to 1.3.21 and this has broken our forms. Whenever we submit a form (AJAX) that passes client side validation, the following error is thrown:

````
Unable to parse response `SyntaxError: Unexpected token < in JSON at position 0`.
````

I haven't been able to find out yet which code is causing this, but I do know the 'bug' is introduced in version 1.3.21. We downgraded Formie to 1.3.19 and 1.3.20 and both work fine, after going back to **1.3.21,** the submission of forms are broken again.

### Form settings

- Multi-page form: No
- Submission Method: Ajax
- Client-side Validation: Yes
- Custom Form Templates: No

**Additional info**
- Plugin version: 1.3.21
- Craft version: 3.6.8
- Multi-site: Yes

**Comment by engram-design** [comment]
There should be further information in the storage/logs/formie.log file, or you can inspect the "Preview" of the request in your network tools in the browser. For instance, here's a similar error (not the same one) and how you can determine more information:

![image](https://user-images.githubusercontent.com/1221575/110029540-b32c9600-7d88-11eb-8f7f-40e7b5c9943a.png)

Usually this occurs if there's a server-side response issue, or the controller didn't return valid JSON, which can happen if the header's aren't correct. This usually only occurs if you're submitting the form with your own JS.

**Comment by engram-design** [comment]
Any further feedback on this?

**Comment by thijskaspers** [comment]
Hey Josh,

I'm struggling with debugging this lol, the formie.log file doesn't show anything and the response preview in the inspector is throwing the raw HTML of the page the contactform is on, instead of a stacktrace to point me in the right direction. The JSON unexpected '<' token error is probably the `<` from the `<!DOCTYPE html>` or `<html>` opening tag.

**Comment by engram-design** [comment]
Yeah, so normally that means that the request isn't being determined as a JS request, and returns a rendered page, instead of a JSON response. Normally, this isn't an issue with our JavaScript that sends the correct headers to the controller endpoint.

Are you able to have a look at the headers for the request being sent when you submit the form?

**Comment by thijskaspers** [comment]
> Hey Josh,
> 
> I'm struggling with debugging this lol, the formie.log file doesn't show anything and the response preview in the inspector is throwing the raw HTML of the page the contactform is on, instead of a stacktrace to point me in the right direction. The JSON unexpected '<' token error is probably the `<` from the `<!DOCTYPE html>` or `<html>` opening tag.

Correction, it is returning the HTML of the homepage, instead of the HTML of the page the contactform is on (/nl/contact).

We are using `craft.formie.renderForm` by the way, and aren't overwriting any templates. Also we aren't using any custom JS, except for the CSRF JS example from the documentation to make Formie work with Blitz.

**Comment by thijskaspers** [comment]
> Yeah, so normally that means that the request isn't being determined as a JS request, and returns a rendered page, instead of a JSON response. Normally, this isn't an issue with our JavaScript that sends the correct headers to the controller endpoint.
> 
> Are you able to have a look at the headers for the request being sent when you submit the form?

These are the headers of the POST request being sent:

````
POST / HTTP/1.1
Host: localhost
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.16; rv:86.0) Gecko/20100101 Firefox/86.0
Accept: application/json
Accept-Language: nl,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate
Referer: http://localhost/nl/contact
X-Requested-With: XMLHttpRequest
Cache-Control: no-cache
Content-Type: multipart/form-data; boundary=---------------------------5384455604092920688120976393
Content-Length: 2023
Origin: http://localhost
Connection: keep-alive
````

And this is the response:

````
HTTP/1.1 301 Moved Permanently
Server: nginx
Date: Tue, 09 Mar 2021 09:15:43 GMT
Content-Type: text/html
Content-Length: 162
Location: http://localhost/nl/
Connection: keep-alive
Strict-Transport-Security: max-age=15768000; includeSubDomains; preload
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer-when-downgrade
````

As far as I can tell the only headers that have changed from 1.3.20 (that one is still working) to 1.3.21/22 is the addition of `Cache-Control: no-cache` to the POST request?

**Comment by engram-design** [comment]
Ah, I think I may know the issue. It's trying to POST to `http://localhost`, but does that resolve to your Craft site? It's responding with a `301 Moved Permanently` redirect, so it makes me think that there's something not quite right.

I changed the destination of the XHR request to be `/` instead of `siteUrl()` which was the case in 1.3.20. This was due to a whole lot of people having issues with cross-origin domains. For instance, they might be accessing the site on `http://www.site.com/contact` and their `siteUrl` is set to `http://site.com/contact`. Therefore, when using XHR from one domain to another, it wasn't being allowed and failed.

Do you have some sort of redirect setup for the `/` of your project?

**Comment by thijskaspers** [comment]
Aaah you are right! It's this commit: https://github.com/verbb/formie/commit/dc94bff045fb8a0bb72acb1dc58fbcf6e9f22274

![Schermafbeelding 2021-03-09 om 10 45 43](https://user-images.githubusercontent.com/1608184/110451610-a54a7e00-80c4-11eb-9f13-7a9f33a83219.png)

Our site is a multi-site setup with 8 languages. We have a 301 redirect from `/(.*)` to `/nl/(.*)`, to enforce that there is always a locale in the URL when visiting the domain, and we don't have any duplicate content on `/` and `/nl` URL's.

Anything we could do to fix this in a nice way without having to remove the redirect?

**Comment by engram-design** [comment]
Gotcha. I'll have a play around with this shortly to see what we can do, but at least I know what the issue is, thanks!

**Comment by engram-design** [comment]
Just pushed a fix for this if you're up for having a test. To get the fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.3.22",
  "...": "..."
}
```

Then run `composer update`.

**Comment by thijskaspers** [comment]
Hey Josh, sorry for the late reply! I've tested the `dev-craft-3` branch and can confirm it's now working, even with the `/` to `/locale` 301 in place 👍

**Comment by engram-design** [comment]
Good to hear! Fixed in [1.3.23](https://github.com/verbb/formie/releases/tag/1.3.23)
